### ISSUE TITLE
Proposal: FlutterClientChannel for handling mobile connection state

### ISSUE THREAD

**Issue by Sub6Resources** [grpc/grpc-dart#223]
There are several issues with grpc-dart currently when trying to use it in the context of a mobile app (see, for instance, #180 as well as a couple issues I've outlined below). grpc-java has introduced an abstract class called `ManagedChannel` (https://github.com/grpc/grpc-java/blob/master/api/src/main/java/io/grpc/ManagedChannel.java) which includes methods for managing the connection (I believe this is equivalent to `ClientChannel` in grpc-dart). They have also added a specific implementation for Android, called `AndroidManagedChannel` that takes in context information and tracks the apps connectivity state across network changes and failures, reconnecting and shutting down connections when necessary to keep the channel running quickly. I propose that we introduce a similar class for those using Flutter (perhaps called `FlutterClientChannel` or `FlutterManagedChannel`) that will help those using Flutter manage the application's connection state more efficiently.

gRPC-dart version I'm using: grpc-dart 2.0.0+1 (Flutter)

## Issues I've seen

In addition to the issues outlined in #180, I've seen several issues when working with gRPC in the context of a Flutter app:

1. ### Idle Timeout
      The default idle timeout is 5 minutes. On mobile devices, the app can be closed, but not dismissed, and then reopened or the app is left sitting open for more than 5 minutes and the channel enters an `idle` state. For many of the devices we've tested with, when a new call is made, the channel re-enters the `connected` state and the call succeeds. For some devices, however, the channel does not re-enter the connected state, but instead throws a gRPC error: `gRPC Error (14, Error making call: Bad state: The http/2 connection is no longer active and can therefore not be used to make new streams.)`. As a workaround for this, we've had to set our `idleTimeout` to a `Duration` of several days. Because this error is device-specific I believe it is likely related to the next issue.
1. ### Network switching
      Any time the network is switched, either from cellular to WiFi or vice versa, or turning off and on mobile data, or if the device sits in an idle state where the wireless connection is temporarily disabled to save power (there are a plethora of other scenarios as well, since network switching tends to be very device specific), the connection is closed, and any rpc calls on an existing channel throw this error: `gRPC Error (14, Error making call: Bad state: The http/2 connection is no longer active and can therefore not be used to make new streams.)`

**Expected result:** gRPC for Dart, specifically Flutter, should be versatile and resilient enough to handle the relatively unstable connections of a mobile platform.

**Actual result:** Whenever any out-of-the-ordinary connectivity state is reached, streams are closed and channels have to be recreated and re-opened manually.

/cc @ryan-berger

**Comment by friddle** [comment]
It seems like my issues #221 
the dart-grpc and dart-http lib didn't listen the socket closed event.

**Comment by friddle** [comment]
@a14n please help us to fix this problem

**Comment by thanhn08** [comment]
is there any workaround or solution for now while waiting for the official fix?

**Comment by sigurdm** [comment]
I think the gRPC Error (14, Error making call: Bad state: The http/2 connection is no longer active and can therefore not be used to make new streams.) issue is the same as
https://github.com/grpc/grpc-dart/issues/228 and should hopefully be helped by: https://github.com/grpc/grpc-dart/pull/231

If we want to make more robustness on mobile we should have some kind of stream of connectionState passed to the channel, that it could behave appropriately. Not sure what is the right behavior of a call when the connection is down though...
That way it would not be flutter specific.

**Comment by friddle** [comment]
@sigurdm thanks!

**Comment by friddle** [comment]
> is there any workaround or solution for now while waiting for the official fix?

temp fix by use this code blow
```
dependency_overrides:
  grpc:
    git:
      url: git@github.com:sigurdm/grpc-dart.git
      ref: check_http2_connection_health
```

**Comment by namdp** [comment]
> > is there any workaround or solution for now while waiting for the official fix?
> 
> temp fix by use this code blow
> 
> ```
> dependency_overrides:
>   grpc:
>     git:
>       url: git@github.com:sigurdm/grpc-dart.git
>       ref: check_http2_connection_health
> ```

unknown revision `check_http2_connection_health`

**Comment by namdp** [comment]
Fixed in #231. Thanks @sigurdm

**Comment by lucaslcode** [comment]
Is this issue closed?
